src/client.rs

```diff
@@ -434,20 +456,164 @@ impl Client {
+    }
+
+    /// Set the maximum lifetime of connections in the pool.
+    pub fn with_max_conn_lifetime(mut self, max_lifetime: Duration) -> Self {
+        self.with_max_lifetime = Some(max_lifetime);
+        self
+    }
+
+    /// Set the socket read timeout for TCP connections.
+    pub fn with_read_timeout(mut self, read_timeout: Duration) -> Self {
+        self.with_read_timeout = Some(read_timeout);
+        self
+    }
+
+    /// Set the socket write timeout for TCP connections.
+    pub fn with_write_timeout(mut self, write_timeout: Duration) -> Self {
+        self.with_write_timeout = Some(write_timeout);
+        self
+    }
+
+    /// Set the connection timeout for TCP connections.
+    pub fn with_connection_timeout(mut self, connection_timeout: Duration) -> Self {
+        self.with_connection_timeout = Some(connection_timeout);
+        self
+    }
+
+    /// Set the hash function for the client.
+    pub fn with_hash_function(mut self, hash_function: fn(&str) -> u64) -> Self {
+        self.with_hash_function = hash_function;
+        self
+    }
+
+    /// Build the client. This will create a connection pool and return a client, or an error if the connection pool could not be created.
+    pub fn build(self) -> Result<Client, MemcacheError> {
+        let urls = self.target.get_urls();
+        let max_size = self.with_max_size;
+        let min_idle = self.with_min_idle;
+        let max_lifetime = self.with_max_lifetime;
+        let timeout = self.with_connection_timeout;
+
+        let mut connections = vec![];
+
+        for url in urls.iter() {
+            let url = Url::parse(url.as_str()).map_err(|e| MemcacheError::BadURL(e.to_string()))?;
+
+            let mut builder = r2d2::Pool::builder()
+                .max_size(max_size)
+                .min_idle(min_idle)
+                .max_lifetime(max_lifetime);
+
+            if let Some(timeout) = timeout {
+                builder = builder.connection_timeout(timeout);
+            }
+
+            let connection = builder
+                .build(ConnectionManager::new(url))
+                .map_err(|e| MemcacheError::PoolError(e))?;
+
+            connections.push(connection);
+        }
+
+        let client = Client {
+            connections,
+            hash_function: self.with_hash_function,
+        };
+
+        client.set_read_timeout(self.with_read_timeout)?;
+        client.set_write_timeout(self.with_write_timeout)?;
+
+        Ok(client)
+    }
+}
+
 #[cfg(test)]
 mod tests {
+    #[test]
+    fn build_client_happy_path() {
+        let client = super::Client::builder()
+            .add_server("memcache://localhost:12345")
+            .build()
+            .unwrap();
+        assert!(client.version().unwrap()[0].1 != "");
+    }
+
+    #[test]
+    fn build_client_bad_url() {
+        let client = super::Client::builder()
+            .add_server("memcache://localhost:12345:")
+            .build();
+        assert!(client.is_err());
+    }
```

Comment:
Yes

Reply:
> The tests `build_client_happy_path` and `build_client_bad_url` effectively verify the basic functionality of the `ClientBuilder`. However, considering the discussion about adding more test cases, it would be beneficial to include tests that cover more scenarios, such as invalid configurations (e.g., setting a negative timeout) or edge cases in the connection parameters. This would ensure the robustness of the `ClientBuilder` implementation.
> 
> Would you like me to help draft additional test cases covering a wider range of scenarios for the `ClientBuilder`?

Yes